### PR TITLE
[SR-2394][ClangImporter] 'returns_twice' funcs unavailable

### PR DESCRIPTION
**Pull request by modocache**
Resolves [SR-2394](https://bugs.swift.org/browse/SR-2394).

The 'returns_twice' attribute is used to denote a function that may return more than one time. Examples include `vfork` and `setjmp`. The Swift compiler, however, cannot ensure definitive initialization when functions such as these are used.

Mark 'returns_twice' functions as unavailable in Swift.

In turn, this removes the need to explicitly mark `vfork` as unavailable on Darwin platforms, since it is now unavailable everywhere.

For code that relies upon `vfork`, `setjmp`, or other 'returns_twice' functions, this is a backwards-incompatible, source-breaking change.

### CODE REVIEW

**Comment by slavapestov** [comment]
You could also make the decks unavailable in Swift 4 mode only by making the availability conditional on language version -- I think we have the plumbing for that in place now.

I can't imagine using setjmp and longjmp is safe in swift -- not only are there issues with DI but destructors and ARC too. But perhaps someone is using it in a narrow case where it happens to work in Swift 3.

File: lib/ClangImporter/ImportDecl.cpp [hunk 1]
```diff
@@ -2953,6 +2953,18 @@ namespace {
       if (decl->isVariadic()) {
         Impl.markUnavailable(result, "Variadic function is unavailable");
       }
+
+      if (decl->hasAttr<clang::ReturnsTwiceAttr>()) {
+        // The Clang 'returns_twice' attribute is used for functions like 'fork'
```
Comment (tkremenek) [inline lib/ClangImporter/ImportDecl.cpp]: Comment says `fork`; should be `vfork`.

File: stdlib/public/Platform/Platform.swift [hunk 2]
```diff
@@ -351,11 +351,6 @@ public func ioctl(
 public func fork() -> Int32 {
   fatalError("unavailable function can't be called")
 }
-
-@available(*, unavailable, message: "Please use threads or posix_spawn*()")
-public func vfork() -> Int32 {
```
Comment (tkremenek) [inline stdlib/public/Platform/Platform.swift]: Stupid question: are we sure `vfork` has the `returns_twice` attribute on all platforms?  It also seems unfortunate to lose this curated availability diagnostic for this API, although reducingthe size of the overlay is good.

Comment (tkremenek) [inline stdlib/public/Platform/Platform.swift]: Ah, forgot that Clang defines `vfork` as a builtin, defining it to have the `returns_twice` attribute.

The loss of the curated diagnostic is still unfortunate, but probably OK.

Comment (modocache) [inline stdlib/public/Platform/Platform.swift]: True, and I wondered about the same. I could undo this change and leave this message in, if you'd prefer.

**Comment by modocache** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
I told Brian I was okay with removing setjmp/longjmp from the equation, and not worrying about it technically being an API change. The best you can get from today's Swift is "it happened not to crash", and pretty much any compiler update could change that.

File: stdlib/public/Platform/Platform.swift (hunk 2 above)
Comment (jrose-apple) [inline stdlib/public/Platform/Platform.swift]: Now that you two have brought it up, I'm inclined to leave it in. It's a small thing that does improve the user experience.

File: lib/ClangImporter/ImportDecl.cpp [hunk 3]
```diff
@@ -2953,6 +2953,18 @@ namespace {
...
+      if (decl->hasAttr<clang::ReturnsTwiceAttr>()) {
+        // The Clang 'returns_twice' attribute is used for functions like
+        // 'vfork' or 'setjmp'. Because these functions may return control flow
+        // of a Swift program to an arbitrary point, Swift's guarantees of
+        // definitive initialization of variables cannot be upheld. As a result,
+        // functions like these cannot be used in Swift.
+        Impl.markUnavailable(
+          result,
+          "Function that may return more than one time (annotated with the "
+          "'returns_twice' attribute) is unavailable");
```
Comment (jrose-apple) [inline lib/ClangImporter/ImportDecl.cpp]: Nitpick: despite the variadic message above, I think it's better to phrase this as a general statement, i.e. *"functions* that may…are unavailable". That makes it clearer that it's not just a one-off rule. It may also be worth adding "in Swift", since that's an important part of it.

**Comment by modocache** [comment]
@swift-ci please smoke test

**Comment by modocache** [comment]
@swift-ci please smoke test

**Comment by modocache** [comment]
@swift-ci please smoke test

**Comment by modocache** [comment]
@swift-ci please smoke test

**Comment by modocache** [comment]
@swift-ci please smoke test

**Comment by modocache** [comment]
This looks ready to go! I left the old diagnostic in. The test got a little gnarlier because of type differences across macOS and Linux.

File: test/ClangImporter/availability_returns_twice.swift [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+#if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
+  import Darwin
+  typealias JumpBuffer = Int32
+#else
+  import Glibc
+  typealias JumpBuffer = __jmp_buf_tag
+#endif
+
+func test_unavailable_returns_twice_function() {
+  var x: JumpBuffer
```
Comment (jrose-apple) [inline test/ClangImporter/availability_returns_twice.swift]: Huh. Using `jmp_buf` directly didn't cut it?

**Review by jrose-apple — APPROVED** [review]

File: test/ClangImporter/availability_returns_twice.swift (hunk 4 above)
Comment (modocache) [inline test/ClangImporter/availability_returns_twice.swift]: Oh, maybe it would have. I noticed the problem from the Linux Swift CI logs, and didn't bother experimenting on my Linux VM. I'll try it out before merging, though.

Comment (modocache) [inline test/ClangImporter/availability_returns_twice.swift]: Unfortunately no, I get the following error on macOS: `cannot convert value of type 'jmp_buf' (aka '(Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32)') to expected argument type 'Int32'`. It works on Linux, though, so I'll use `jmp_buf` there instead of `__jmp_buf_tag`.

**Comment by modocache** [comment]
@swift-ci please smoke test and merge

**Comment by modocache** [comment]
@swift-ci please smoke test and merge

File: test/ClangImporter/availability_returns_twice.swift (hunk 4 above)
Comment (jrose-apple) [inline test/ClangImporter/availability_returns_twice.swift]: Ah, okay. Fixed-sized arrays ruin everything.

Comment (modocache) [inline test/ClangImporter/availability_returns_twice.swift]: Can support for these be improved somehow? Alternatively, is there a good discussion on why this is imported weirdly on macOS?

Comment (jrose-apple) [inline test/ClangImporter/availability_returns_twice.swift]: The particular improvement we could make is to allow `&` applied to a homogeneous tuple (today's fixed-sized array) implicitly convert to a pointer to the first element, without having to go through type rebinding. cc @atrick

**Comment by atrick** [comment]
> On Jan 9, 2017, at 10:21 AM, Jordan Rose <notifications@github.com> wrote:
> 
> @jrose-apple commented on this pull request.
> 
> In test/ClangImporter/availability_returns_twice.swift <https://github.com/apple/swift/pull/6513>:
> 
> > @@ -0,0 +1,15 @@
> +// RUN: %target-swift-frontend -typecheck -verify %s
> +
> +#if os(OSX) || os(iOS) || os(watchOS) || os(tvOS)
> +  import Darwin
> +  typealias JumpBuffer = Int32
> +#else
> +  import Glibc
> +  typealias JumpBuffer = __jmp_buf_tag
> +#endif
> +
> +func test_unavailable_returns_twice_function() {
> +  var x: JumpBuffer
> The particular improvement we could make is to allow & applied to a homogeneous tuple (today's fixed-sized array) implicitly convert to a pointer to the first element, without having to go through type rebinding. cc @atrick <https://github.com/atrick>
This is a real usability problem. I filed this a while back on the diagnostics:
https://bugs.swift.org/browse/SR-2471 <https://bugs.swift.org/browse/SR-2471>

And I agree we should just support implicit casting in this case:
https://bugs.swift.org/browse/SR-3590

-Andy

**Comment by slavapestov** [comment]
@modocache Do you mind creating a new PR to add an entry to CHANGLEOG.md as well? We have several new ClangImporter features that might affect users so I'm following up with PR authors about documenting them.

**Comment by modocache** [comment]
Will do! Thanks.

**Comment by modocache** [comment]
Here's a pull request for the CHANGELOG entry: https://github.com/apple/swift/pull/6733
